## Customer Churn Prediction and Analysis

All data about the dataset used in this project is available [here](https://www.kaggle.com/datasets/anandshaw2001/customer-churn-dataset)

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
data = pd.read_csv(r"C:\Users\youss\OneDrive\Desktop\DataQuest\Projects\Customer Churn\Churn_Modelling.csv")

In [3]:
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [5]:
data.isna().sum()

RowNumber          0
CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

In [6]:
data.describe()

RowNumber    CustomerId   CreditScore           Age        Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  10000.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800      5.012800   
std     2886.89568  7.193619e+04     96.653299     10.487806      2.892174   
min        1.00000  1.556570e+07    350.000000     18.000000      0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000      3.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000      5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000      7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000     10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

In [7]:
churn = data.copy()

In [8]:
churn['Geography'].unique()

array(['France', 'Spain', 'Germany'], dtype=object)

**Country with the most customer churns**

In [9]:
exits = churn[churn['Exited'] == 1][['Geography', 'Exited']]
count = exits['Geography'].value_counts()
count

Geography
Germany    814
France     810
Spain      413
Name: count, dtype: int64

**Gender Percentage in the Dataset**

In [10]:
churn['Gender'].value_counts(normalize=True)*100

Gender
Male      54.57
Female    45.43
Name: proportion, dtype: float64

**Most Churned Age Category with Credit Card Holders**

In [11]:
age = churn.copy()
age.loc[(age['Age'] >= 18) & (age['Age'] < 43), 'age_category'] = 'Young Adults'
age.loc[(age['Age'] >= 43) & (age['Age'] < 68), 'age_category'] = 'Middle-Aged Adults'
age.loc[(age['Age'] >= 68) & (age['Age'] < 93), 'age_category'] = 'Senior Adults'

age = age[(age['HasCrCard'] == 1) & (age['Exited'] == 1)]

In [12]:
round(age['age_category'].value_counts(normalize=True)*100, 0)

age_category
Middle-Aged Adults    57.0
Young Adults          42.0
Senior Adults          1.0
Name: proportion, dtype: float64

In [13]:
churn['Geography'] = churn['Geography'].map({'France':1, 'Germany':2, 'Spain':3})
churn['Gender'] = churn['Gender'].map({'Male':0, 'Female':1})

In [14]:
churn['Exited'].value_counts()

Exited
0    7963
1    2037
Name: count, dtype: int64

We can see an obvious inbalance between the two classes of the target column, if we trained a model on this dataset we will have a biased model, to fix that we will take a balanced sample from this dataset and train the model on the sample.

In [15]:
non_churn = churn[churn['Exited'] == 0]
churnn = churn[churn['Exited'] == 1]

min_class_size = min(len(non_churn), len(churnn))

non_churn_sample = non_churn.sample(n=min_class_size, random_state=42)
churn_sample = churnn.sample(n=min_class_size, random_state=42)

balanced_sample = pd.concat([non_churn_sample, churn_sample])

balanced_sample = balanced_sample.sample(frac=1, random_state=42).reset_index(drop=True)

print(balanced_sample['Exited'].value_counts())

Exited
0    2037
1    2037
Name: count, dtype: int64


In [16]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
import numpy as np

balanced_sample = balanced_sample.drop(columns=['RowNumber', 'CustomerId', 'Surname'])
balanced_sample = balanced_sample.fillna(balanced_sample.mean())
X = balanced_sample.drop(columns=['Exited']) 
y = balanced_sample['Exited'] 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

logreg = LogisticRegression()
logreg.fit(X_train, y_train)

coefficients = logreg.coef_[0]
feature_names = X.columns

odds_ratios = np.exp(coefficients)
for feature, odds in zip(feature_names, odds_ratios):
    print(f'{feature}: {odds}')

CreditScore: 0.9961832801263591
Geography: 1.0004841884866869
Gender: 1.0009744208906117
Age: 1.0567034620047415
Tenure: 0.9995535373847559
Balance: 1.000004678751691
NumOfProducts: 0.9995788745577661
HasCrCard: 0.9999942174851191
IsActiveMember: 0.9984379598158026
EstimatedSalary: 0.9999989093309617


- `Age` has the most significant impact, with an odds ratio of 1.056, meaning older individuals are more likely to hit the target outcome.

- Features like `CreditScore`, `Tenure`, and `IsActiveMember` have small effects, but they still slightly influence the odds. 

- Features such as `Geography`, `Gender`, `Balance`, `NumOfProducts`, `HasCrCard`, and `EstimatedSalary` have odds ratios very close to 1, meaning they have minimal to no practical effect on the target outcome.

In [17]:
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, roc_auc_score
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

X = balanced_sample.drop(columns=['Exited', 'Gender', 'HasCrCard', 'Balance']) 
y = balanced_sample['Exited']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

rf = RandomForestClassifier(
    n_estimators=200,       
    max_depth=30,           
    min_samples_split=5,    
    min_samples_leaf=2,     
    max_features='sqrt',    
    bootstrap=True,         
    criterion='entropy',       
    class_weight='balanced', 
    random_state=42         
)

rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)

accuracy_rf = accuracy_score(y_test, y_pred_rf)
print("Random Forest Accuracy:", accuracy_rf)
print("Random Forest Confusion Matrix:\n", confusion_matrix(y_test, y_pred_rf))
print("Random Forest Classification Report:\n", classification_report(y_test, y_pred_rf))

roc_auc_rf = roc_auc_score(y_test, rf.predict_proba(X_test)[:, 1])
print("Random Forest ROC-AUC Score:", roc_auc_rf)

comparison_df = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred_rf})
print(comparison_df.head(10))


Random Forest Accuracy: 0.7522485690923958
Random Forest Confusion Matrix:
 [[474 137]
 [166 446]]
Random Forest Classification Report:
               precision    recall  f1-score   support

           0       0.74      0.78      0.76       611
           1       0.77      0.73      0.75       612

    accuracy                           0.75      1223
   macro avg       0.75      0.75      0.75      1223
weighted avg       0.75      0.75      0.75      1223

Random Forest ROC-AUC Score: 0.8322689686894944
      Actual  Predicted
1749       0          0
2053       0          0
538        1          1
438        0          0
2683       0          1
402        1          0
1027       0          0
308        1          0
2437       1          0
149        0          0


## Final Outcome

**Accuracy (0.7522):**

- The accuracy of the model is 75.22%, meaning that 75.22% of the predictions made by the Random Forest model are correct.

**Confusion Matrix:**

- 474: True Negatives (TN) – Correctly predicted as class 0.

- 137: False Positives (FP) – Incorrectly predicted as class 1, but actually class 0.

- 166: False Negatives (FN) – Incorrectly predicted as class 0, but actually class 1.

- 446: True Positives (TP) – Correctly predicted as class 1.

**Classification Report:**

**Precision: The proportion of true positives (correctly predicted) among all positive predictions.**

- Class 0: 74% of instances predicted as class 0 are correct.

- Class 1: 77% of instances predicted as class 1 are correct.

**Recall: The proportion of true positives among all actual positives.**

- Class 0: 78% of class 0 instances are correctly identified.

- Class 1: 73% of class 1 instances are correctly identified.

**F1-Score: Harmonic mean of precision and recall.**

- Class 0: 76%

- Class 1: 75%

**Support: The number of actual occurrences for each class.**

- Class 0: 611

- Class 1: 612

**ROC-AUC Score (0.8323):**

- The ROC-AUC score is 0.8323, meaning the model's ability to distinguish between the two classes is 83.23%. An ROC-AUC score of 1 represents a perfect model, while 0.5 represents random guessing.